### PR TITLE
feat: Enhance filter expression parsing and evaluation

### PR DESCRIPTION
**Pull request by isSerge**
# Summary

This PR is an attempt to address issues discussed in #204

**Key changes:**

Introduction of `Expression` sub-module responsible for:
- AST definitions
- parsing (using `winnow` crate)
- shared evaluation logic (`ConditionEvaluator` trait, evaluation errors, helpers)

EVM and Stellar both have implementations of `ConditionEvaluator` (`evaluator.rs`) responsible for:
- conversion of parameter values (LHS) into chain-specific types
- LHS vs RHS comparison

`evaluate_expression` method in chain filters now only delegates parsing and evaluation to dedicated modules

**Breaking changes:**
- Expression parsing is more strict due to parsing and evaluation decoupling. For example in expression`"input contains '567890'"` RHS now **has** to be in single quotes to let parser now it's a string, otherwise it result in type mismatch error
- Stellar syntax `arguments[0][1] == 20` is no longer valid. LHS has to be an _identifier_ (can introduce keyword if this is crucial)

**Next steps:**
- Refactor `evaluate_expression` to return `Result<bool, EvaluationError>`
- Parsing of implicit arrays for EVM (e.g. `uint8` `"1,2,3"`)
- Stellar events key access `0.2`

If any of these is critical - let me know and I will include here. Otherwise I think there are already a lot of changes to review.

I haven't updated Antora docs yet, will add a commit once we discuss and confirm implementation. 

## Checklist

- [ ] Add a reference to related issues in the PR description.
- [ ] Add unit tests if applicable.
- [ ] Add integration tests if applicable.
- [ ] Add property-based tests if applicable.
- [ ] Update documentation if applicable.

### CODE REVIEW

File: src/services/filter/filters/evm/filter.rs [hunk 1]
```diff
@@ -411,123 +413,34 @@ impl<T> EVMBlockFilter<T> {
 		expression: &str,
 		args: &Option<Vec<EVMMatchParamEntry>>,
 	) -> bool {
+		// TODO: double-check cases where args is None
```
Comment (shahnami) [inline src/services/filter/filters/evm/filter.rs:416]: Let's not forget this

File: src/services/filter/filters/evm/filter.rs [hunk 2]
```diff
@@ -2181,50 +2142,103 @@ mod tests {
...
+
+		// Test object in array
+		assert!(filter.evaluate_expression("objects[0].name == 'Alice'", &args));
+		assert!(filter.evaluate_expression("objects[1].name == 'Bob'", &args));
+
+		// Test out-of-bounds access
+		assert!(!filter.evaluate_expression("objects[2].name == 'Charlie'", &args));
+	}
+
+	// TODO: replace with actual error tests when method returns Result
```
Comment (shahnami) [inline src/services/filter/filters/evm/filter.rs:2241]: Let's not forget this one

**Review by shahnami — COMMENTED** [review]
The code looks solid overall! We've got good test coverage and the error handling is pretty robust. 

There are a few things we should clean up though. We've got some TODOs hanging around we should probably tackle those first to keep things tidy. 

The tests are good, but we could make them even better by adding property-based testing. This would help us catch edge cases with different input sizes and make sure we're covering all our types properly.

As for `EvaluationError` we should probably add a helper implementation similar to what we did with `NotificationError`,  it'll make the error handling more consistent and give us better logging. 

Documentation-wise, we need to beef up both the Rust docs (for the devs) and the Antora docs (for the users). The Rust docs should be more technical and detailed, while Antora should focus on how to actually use the thing.

Lastly, more of a nit, but we might want to think about adding some input validation limits (like max array size, expression length, etc.). It's not super critical, but it could help with performance and prevent any potential resource issues. We should check if winnow has anything built-in for this.

Thank you for your hard work on this!

I'm tagging @NicoMolinaOZ and @KitHat for additional reviews on this given the size and complexity (they are OOO, so could be by EOM).

File: src/services/filter/filters/evm/evaluator.rs [hunk 3]
```diff
@@ -0,0 +1,1146 @@
...
+			"address" => self.compare_address(lhs_value_str, operator, rhs_literal),
+			"string" | "bytes" | "bytes32" => {
+				self.compare_string(lhs_value_str, operator, rhs_literal)
+			}
+			"bool" => self.compare_boolean(lhs_value_str, operator, rhs_literal),
+			_ => Err(EvaluationError::TypeMismatch(format!(
+				"Unsupported EVM parameter kind for comparison: {}",
+				lhs_kind_str
+			))),
+		}
```
[the remark this answers is not in the document]
Comment (isSerge) [reply src/services/filter/filters/evm/evaluator.rs:303]: I was planning to add these in the following PR (mentioned in the **next steps** section above), but if these are critical - I can add here

[the remark this answers is not in the document]
Comment (isSerge) [reply src/services/filter/filters/evm/evaluator.rs:303]: same for objects

File: src/services/filter/filters/evm/filter.rs [hunk 4]
```diff
@@ -398,6 +399,7 @@ impl<T> EVMBlockFilter<T> {
 		}
 	}
 
+	// TODO: should return Result<bool>
```
[the remark this answers is not in the document]
Comment (isSerge) [reply src/services/filter/filters/evm/filter.rs:402]: yes, I was planning to do this in a separate PR, because this change requires refactoring tests and there were some challenges when I tried it initially

File: src/services/filter/filters/evm/evaluator.rs [hunk 5]
```diff
@@ -0,0 +1,1146 @@
...
+
+	// Helper to create a dummy EVMConditionEvaluator (args don't matter for these unit tests)
+	fn create_evaluator() -> EVMConditionEvaluator<'static> {
+		static EMPTY_ARGS: EVMArgs = Vec::new();
+		EVMConditionEvaluator::new(&EMPTY_ARGS)
+	}
+
+	/// --- Test cases for compare_u256 ---
+	#[test]
+	fn test_compare_u256_valid() {
```
[the remark this answers is not in the document]
Comment (isSerge) [reply src/services/filter/filters/evm/evaluator.rs:375]: It fails because RHS value is out of range (`U256::MAX` + 1), which is expected. 
I will add some unit tests for `string_to_u256` and `string_to_i256` anyway

**Comment by isSerge** [comment]
@shahnami regarding this:
> we might want to think about adding some input validation limits (like max array size, expression length, etc.). It's not super critical, but it could help with performance and prevent any potential resource issues. We should check if winnow has anything built-in for this.

`winnow` does not have built-in features for this, but we can enforce these limits in parser functions. I can address this later if you could create an issue with some specific limits

File: src/services/filter/filters/stellar/filter.rs [hunk 6]
```diff
@@ -2608,988 +2161,501 @@ mod tests {
...
-		// Test greater than or equal operator
-		assert!(!filter.compare_bool("true", ">=", "false"));
-
-		// Test less than or equal operator
-		assert!(!filter.compare_bool("false", "<=", "true"));
-
-		// Test empty operator
-		assert!(!filter.compare_bool("true", "", "false"));
+		// Test true conditions
+		assert!(filter.evaluate_expression("array_str[0] == '100'", &args));
```
[the remark this answers is not in the document]
Comment (isSerge) [reply src/services/filter/filters/stellar/filter.rs:2288]: added

File: src/services/filter/filters/evm/filter.rs [hunk 7]
```diff
@@ -2181,50 +2142,103 @@ mod tests {
...
 
+	#[test]
+	fn test_evaluate_expression_array_indexing() {
+		let filter = create_test_filter();
+		let args = Some(vec![create_test_param("array", "[1, 2, 3]", "array")]);
+
+		// Test array indexing
+		assert!(filter.evaluate_expression("array[0] == 1", &args));
+		assert!(filter.evaluate_expression("array[1] == 2", &args));
+		assert!(filter.evaluate_expression("array[2] == 3", &args));
```
[the remark this answers is not in the document]
Comment (isSerge) [reply src/services/filter/filters/evm/filter.rs:2243]: added

File: src/services/filter/filters/stellar/filter.rs [hunk 8]
```diff
@@ -2608,988 +2161,501 @@ mod tests {
...
+		assert!(filter.evaluate_expression("nested_object.outer.inner == 'value'", &args));
+		assert!(filter.evaluate_expression("nested_object.outer.inner != 'wrong'", &args));
 
-		// Test empty strings
-		assert!(!filter.compare_bool("", "==", "true"));
-		assert!(!filter.compare_bool("true", "==", ""));
-		assert!(!filter.compare_bool("", "==", ""));
+		// Test false conditions
+		assert!(!filter.evaluate_expression("nested_object.outer.inner == 'wrong'", &args));
+		assert!(!filter.evaluate_expression("nested_object.outer.inner != 'value'", &args));
```
[the remark this answers is not in the document]
Comment (isSerge) [reply src/services/filter/filters/stellar/filter.rs:2266]: Should I create a separate issue to address this later? I have included full comparison for arrays, but objects seem less critical

File: src/services/filter/filters/evm/filter.rs (hunk 1 above)
Comment (shahnami) [reply src/services/filter/filters/evm/filter.rs:422]: Yes, let's refactor it with the rest in a separate PR 👍

File: src/services/filter/expression/ast.rs
Comment (NicoMolinaOZ) [inline src/services/filter/expression/ast.rs:1]: Very nice improvement implementing AST logic, thanks man!

File: src/services/filter/expression/helpers.rs [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+/// Traverses the Expression AST and uses ConditionEvaluator to evaluate conditions
+/// Returns true if the expression evaluates to true, false otherwise
+/// Returns an error if the evaluation fails
+pub fn evaluate(
+	expression: &Expression<'_>,
+	evaluator: &impl ConditionEvaluator,
+) -> Result<bool, EvaluationError> {
+	match expression {
+		Expression::Condition(condition) => {
+			println!("{:?}", condition.left);
```
[the remark this answers is not in the document]
Comment (isSerge) [reply src/services/filter/expression/helpers.rs]: shame on me

File: tests/properties/filters/evm/filter.rs
Comment (NicoMolinaOZ) [inline tests/properties/filters/evm/filter.rs:1]: Thanks for this PBT, looks good!

**Review by NicoMolinaOZ — APPROVED** [review]
LGTM! thanks for the effort, looks very nice, lot of improvements 👏

**Comment by isSerge** [comment]
@shahnami there is a failing test unrelated to this PR blocking CI

**Comment by shahnami** [comment]
> @shahnami there is a failing test unrelated to this PR blocking CI

Due to some race conditions the JS related test can be flaky at times, we'll investigate at some point, but a re-run usually does the trick (for now). I just re-ran and your tests passed. However, I see docker-scan failed, but expecting @tirumerla pending PR to fix that.

**Comment by shahnami** [comment]
Issue should now be fixed in main branch. You can pull it in and it should pass all tests.

**Comment by shahnami** [comment]
@isSerge I'm thinking would it would make sense to start the work on the follow-up PRs and review them individually but merge them all into this one before we merge into main?

We're mainly just waiting for:

- Refactor evaluate_expression to return Result<bool, EvaluationError>
- Parsing of implicit arrays for EVM (e.g. uint8[] "1,2,3")
- Stellar events key access 0.2

However, I think the latter you've already integrated, right?

**Comment by isSerge** [comment]
@shahnami yes, Stellar key access is already there. Let me submit the rest

**Comment by isSerge** [comment]
@shahnami the thing is these are all implemented in my forked repo branches. I can consolidate all changes within my fork and include here, or can submit separate stacked PRs to the `main`. Which do you prefer?

**Comment by shahnami** [comment]
> @shahnami the thing is these are all implemented in my forked repo branches. I can consolidate all changes within my fork and include here, or can submit separate stacked PRs to the `main`. Which do you prefer?

I see, okay, let's add them to this PR then

**Comment by isSerge** [comment]
@shahnami I have included changes to handle array types, direct object (map) comparisons and return `Result` for `evaluate_expression` methods. 

There are some integration tests failing occasionally:
`integration::bootstrap::main::test_create_trigger_handler`
`integration::bootstrap::main::test_create_trigger_handler_empty_matches`
`integration::bootstrap::main::test_create_trigger_handler_with_conditions`
`integration::blockwatcher::service::test_max_past_blocks_limit`
`integration::blockwatcher::service::test_fresh_start_processing`
but seems unrelated to my changes

**Review by shahnami — APPROVED** [review]

**Comment by isSerge** [comment]
@NicoMolinaOZ I suspect the problem is this expression: `2 contains '1300000,1293965'`: `contains` operator only checks if a single item is present in LHS, in this case RHS is parsed as a string `'1300000,1293965'`

**Comment by NicoMolinaOZ** [comment]
> @NicoMolinaOZ I suspect the problem is this expression: `2 contains '1300000,1293965'`: `contains` operator only checks if a single item is present in LHS, in this case RHS is parsed as a string `'1300000,1293965'`

Thanks for the response. I saw that we are having some issues parsing complex types for EVM functions and events. When you have an object like this:
`data {
 field: string[]
 another_field: number
}` solidity under the hood converts it to a `tuple`, and actually, we don't support parsing tuples. We want to merge this PR and implement these improvements in the future.

Thanks!
